### PR TITLE
Improve contention handling of persistent cache locks

### PR DESCRIPTION
**Pull request by jjohannes**
The contention handling of lock of mode `Exclusive` and `None` (release on request) can become inefficient if the process holding a lock becomes unresponsive/slow (e.g. a daemon under memory pressure) or if many processes fight for the same lock.

This PR improves this by:
1. **Requesting Side** - Do not send hundreds of similar lock release request over the socket if a lock release takes longer
2. **Receiving Side** - Confirm a request to the receiver to support (1.)
3. **Receiving Side** - Do not block the socket listener thread by executing the _lock release action_ on a separate thread
4. **Requesting Side** -In case several processes fight for the lock: reset timeout if the lock owner changes

### CODE REVIEW

File: subprojects/core/src/testFixtures/groovy/org/gradle/cache/internal/locklistener/NoOpFileLockContentionHandler.java [hunk 1]
```diff
@@ -26,6 +26,7 @@ public int reservePort() {
         return -1;
     }
 
-    public void pingOwner(int port, long lockId, String displayName) {
+    public boolean maybePingOwner(int port, long lockId, String displayName, long timeElapsed) {
```
Comment (oehme) [inline subprojects/core/src/testFixtures/groovy/org/gradle/cache/internal/locklistener/NoOpFileLockContentionHandler.java]: ⭕️ Shouldn't the elapsed time since the last ping be handled in the handler instead of passed in?

File: subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/DefaultFileLockContentionHandler.java [hunk 2]
```diff
@@ -23,20 +23,28 @@
...
 import java.util.Map;
 import java.util.concurrent.locks.Lock;
 import java.util.concurrent.locks.ReentrantLock;
 
 public class DefaultFileLockContentionHandler implements FileLockContentionHandler, Stoppable {
     private static final Logger LOGGER = LoggerFactory.getLogger(DefaultFileLockContentionHandler.class);
+    private static final int PING_DELAY = 1000;
     private final Lock lock = new ReentrantLock();
     private final Map<Long, Runnable> contendedActions = new HashMap<Long, Runnable>();
+    private final List<Long> contendedActionsRunning = new ArrayList<Long>();
```
Comment (oehme) [inline subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/DefaultFileLockContentionHandler.java]: ❌ All these added data structures scream "There's a new class waiting for extraction".

File: subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/DefaultFileLockContentionHandler.java [hunk 3]
```diff
@@ -23,20 +23,28 @@
...
 
+import java.net.DatagramPacket;
+import java.util.ArrayList;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 import java.util.concurrent.locks.Lock;
 import java.util.concurrent.locks.ReentrantLock;
 
 public class DefaultFileLockContentionHandler implements FileLockContentionHandler, Stoppable {
```
Comment (oehme) [inline subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/DefaultFileLockContentionHandler.java]: ❌ This really needs some documentation for the next person trying to understand this code.

File: subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/DefaultFileLockContentionHandler.java [hunk 4]
```diff
@@ -60,39 +68,57 @@ public void run() {
...
+                            contendedActionsRunning.add(lockId);
                         }
-                    } finally {
-                        lock.unlock();
+
+                    }
+                    lock.unlock();
+
+                    if (action != null) {
+                        // I have the action, I execute it and tell the other side that I am on it
```
Comment (oehme) [inline subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/DefaultFileLockContentionHandler.java]: What does "I have the action" mean?

File: subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/DefaultFileLockContentionHandler.java [hunk 5]
```diff
@@ -60,39 +68,57 @@ public void run() {
...
+                        lockId = communicator.decodeLockId(packet);
                     } catch (GracefullyStoppedException e) {
                         return;
                     }
-                    lock.lock();
                     Runnable action;
-                    try {
+
+                    lock.lock();
+                    if (contendedActionsRunning.contains(lockId)) {
```
Comment (oehme) [inline subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/DefaultFileLockContentionHandler.java]: I don't understand this. The action is running, so I set it to null and do nothing further down? What does that mean?

The code is very hard to follow. I suggest extracting a State class with all these maps and giving that class easy to understand methods.

File: subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/FileLockCommunicator.java [hunk 6]
```diff
@@ -68,14 +69,39 @@ public void pingOwner(int ownerPort, long lockId, String displayName) {
...
+            return packet;
+        } catch (IOException e) {
+            if (!stopped) {
+                throw new RuntimeException(e);
+            }
+            throw new GracefullyStoppedException();
+        }
+    }
+
+    public long decodeLockId(DatagramPacket receivedPacket) {
```
Comment (oehme) [inline subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/FileLockCommunicator.java]: ⭕️ Why was this split up? Where do we care about the packet itself?

File: subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/DefaultFileLockManager.java [hunk 7]
```diff
@@ -146,6 +145,9 @@ public DefaultFileLock(File target, LockOptions options, String displayName, Str
             LockStateSerializer stateProtocol = options.isUseCrossVersionImplementation() ? new Version1LockStateSerializer() : new DefaultLockStateSerializer();
             lockFileAccess = new LockFileAccess(lockFile, new LockStateAccess(stateProtocol));
             try {
+                if (whenContended != null) {
```
Comment (oehme) [inline subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/DefaultFileLockManager.java]: ⭕️ What does this change fix? I only see adjusted mock tests in this commit, but no integration test (maybe that comes later). Can you just point me to the test coverage?

File: subprojects/persistent-cache/src/integTest/groovy/org/gradle/cache/internal/DefaultFileLockManagerContentionIntegrationTest.groovy [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+                    if (selectedAddress == null) {
+                        selectedAddress = packet.address
+                    }
+                    if (selectedAddress == packet.address) {
+                        reactOnRequest.run()
+                    }
+                }
+
+            }
+        }.start()
```
Comment (oehme) [inline subprojects/persistent-cache/src/integTest/groovy/org/gradle/cache/internal/DefaultFileLockManagerContentionIntegrationTest.groovy]: ❌ Should be shut down after the test.

File: subprojects/persistent-cache/src/integTest/groovy/org/gradle/cache/internal/DefaultFileLockManagerContentionIntegrationTest.groovy [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+import org.gradle.internal.remote.internal.inet.InetAddressFactory
+import org.gradle.internal.time.TrueTimeProvider
+
+import java.util.concurrent.Executors
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.TimeoutException
+
+import static org.gradle.test.fixtures.ConcurrentTestUtil.poll
+
+class DefaultFileLockManagerContentionIntegrationTest extends AbstractIntegrationSpec {
```
Comment (oehme) [inline subprojects/persistent-cache/src/integTest/groovy/org/gradle/cache/internal/DefaultFileLockManagerContentionIntegrationTest.groovy]: Did you run this specific test a couple times on CI to ensure it is stable?

File: subprojects/core/src/testFixtures/groovy/org/gradle/cache/internal/locklistener/NoOpFileLockContentionHandler.java (hunk 1 above)
Comment (jjohannes) [inline subprojects/core/src/testFixtures/groovy/org/gradle/cache/internal/locklistener/NoOpFileLockContentionHandler.java]: I was thinking about that. But I didn't want to add another state that depends on time here. I think that adds more complexity.

File: subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/DefaultFileLockContentionHandler.java (hunk 2 above)
Comment (jjohannes) [inline subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/DefaultFileLockContentionHandler.java]: Agreed. I will add Javadoc.

File: subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/FileLockCommunicator.java (hunk 6 above)
Comment (jjohannes) [inline subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/FileLockCommunicator.java]: We need more information than just the decoded lock id in order to confirm ("ping back") to the requester (encoded ID, address, port).
It's then used to call `confirmUnlockRequest`: 
https://github.com/gradle/gradle/blob/5254a577d74ed5caa4e8a5d93c56d9900d5e4b2c/subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/FileLockCommunicator.java#L100-L100

File: subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/DefaultFileLockManager.java (hunk 7 above)
Comment (jjohannes) [inline subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/DefaultFileLockManager.java]: Yes the tests are all added in a different commit. Without this change, the _whenContended_ action is only registered after the lock is acquired. And this means that the lock does not know that it is _allowed to be contended_ and thus will not listen to confirmation messages. 

This test fails without this:
https://github.com/gradle/gradle/blob/0913232e8536330b6087ba0f6f90f5edc6ebc600/subprojects/persistent-cache/src/integTest/groovy/org/gradle/cache/internal/DefaultFileLockManagerContentionIntegrationTest.groovy#L98-L117

File: subprojects/persistent-cache/src/integTest/groovy/org/gradle/cache/internal/DefaultFileLockManagerContentionIntegrationTest.groovy (hunk 8 above)
Comment (jjohannes) [inline subprojects/persistent-cache/src/integTest/groovy/org/gradle/cache/internal/DefaultFileLockManagerContentionIntegrationTest.groovy]: I added cleanup code.

File: subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/DefaultFileLockContentionHandler.java (hunk 5 above)
Comment (jjohannes) [inline subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/DefaultFileLockContentionHandler.java]: I already improved the structure in a follow up commit.

File: subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/DefaultFileLockContentionHandler.java (hunk 3 above)
Comment (jjohannes) [inline subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/DefaultFileLockContentionHandler.java]: Agreed. I will add Javadoc.

File: subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/DefaultFileLockContentionHandler.java (hunk 2 above)
Comment (jjohannes) [inline subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/DefaultFileLockContentionHandler.java]: I put the contended actions state into one structure.

I left `unlocksRequestedFrom` and `unlocksConfirmedFrom` as simple maps from `lockId` to `port` since they only record the incoming messages as they are which are interpreted later in `maybePingOwner`. It's also due to the low level protocol. The packets we send over the socket only contain the lockId and no further context. But I do not want to touch this protocol if not absolutely necessary.

File: subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/DefaultFileLockContentionHandler.java (hunk 4 above)
Comment (jjohannes) [inline subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/DefaultFileLockContentionHandler.java]: Improved this a bit.

File: subprojects/persistent-cache/src/integTest/groovy/org/gradle/cache/internal/DefaultFileLockManagerContentionIntegrationTest.groovy (hunk 9 above)
Comment (jjohannes) [inline subprojects/persistent-cache/src/integTest/groovy/org/gradle/cache/internal/DefaultFileLockManagerContentionIntegrationTest.groovy]: https://builds.gradle.org/project.html?projectId=Gradle_Util&branch_Gradle_Util=jj%2Flocking%2Fimprove-contention-handling

File: subprojects/persistent-cache/src/integTest/groovy/org/gradle/cache/internal/DefaultFileLockManagerContentionIntegrationTest.groovy [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+        pingRequestCount == 3 || pingRequestCount == 4
+        timeProvider.currentTime - startTime > 3000 // See: DefaultFileLockContentionHandler.PING_DELAY
+    }
+
+    def "the lock holder starts the release request only once and discards additional requests in the meantime"() {
+        given:
+        def requestReceived = false
+        setupLockOwner {
+            requestReceived = true
+            while(true) { Thread.sleep(100000) } //block the release action thread
```
Comment (melix) [inline subprojects/persistent-cache/src/integTest/groovy/org/gradle/cache/internal/DefaultFileLockManagerContentionIntegrationTest.groovy]: When is this interrupted?

File: subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/DefaultFileLockContentionHandler.java [hunk 11]
```diff
@@ -60,51 +102,83 @@ public void run() {
...
-                    action.run();
+                    lock.unlock();
                 }
             }
         };
     }
 
+    private void startLockReleaseAsLockHolder(ContendedAction contendedAction) {
+        unlockActionExecutor.execute(contendedAction.action);
+        contendedAction.running = true;
```
Comment (melix) [inline subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/DefaultFileLockContentionHandler.java]: I understand `unlockActionExecutor` will do it asynchronously, but maybe putting `running=true` before would be less puzzling.

File: subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/DefaultFileLockContentionHandler.java [hunk 12]
```diff
@@ -60,51 +102,83 @@ public void run() {
...
+            contendedActions.put(lockId, new ContendedAction(whenContended));
         } finally {
             lock.unlock();
         }
     }
 
-    public void pingOwner(int port, long lockId, String displayName) {
-        getCommunicator().pingOwner(port, lockId, displayName);
+    public boolean maybePingOwner(int port, long lockId, String displayName, long timeElapsed) {
+        if (unlocksConfirmedFrom.containsKey(lockId) && unlocksConfirmedFrom.get(lockId) == port) {
```
Comment (melix) [inline subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/DefaultFileLockContentionHandler.java]: This can be simplified to `unlocksConfirmedPort.get(lockId)==port`

Comment (jjohannes) [inline subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/DefaultFileLockContentionHandler.java]: That throws a NPE if the entry does not exist, because of the comparison with an `int` it tries to unbox the null. I now simplified it like this:
`Integer.valueOf(port).equals(unlocksConfirmedFrom.get(lockId))`

File: subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/DefaultFileLockContentionHandler.java (hunk 11 above)
Comment (jjohannes) [inline subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/locklistener/DefaultFileLockContentionHandler.java]: Agreed.

File: subprojects/persistent-cache/src/integTest/groovy/org/gradle/cache/internal/DefaultFileLockManagerContentionIntegrationTest.groovy (hunk 10 above)
Comment (jjohannes) [inline subprojects/persistent-cache/src/integTest/groovy/org/gradle/cache/internal/DefaultFileLockManagerContentionIntegrationTest.groovy]: When the test worker stops :)? Good catch. Added a clean up to the end of the test.

**Comment by jjohannes** [comment]
As agree with @melix, I am merging this now. We can still keep the discussion going and improve documentation afterwards.
